File: src/core/esoui.py
"""MMOUI/ESOUI API client — clean JSON API, no HTML scraping."""

from __future__ import annotations

import re
import time
from dataclasses import dataclass, field
from typing import Optional

import requests


# ── API endpoints ─────────────────────────────────────────────────────────────

_GLOBAL_CONFIG_URL = "https://api.mmoui.com/v3/globalconfig.json"
_GAME_ID = "ESO"

_SESSION = requests.Session()
_SESSION.headers.update({
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
        "AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36"
    ),
    "Accept-Encoding": "gzip, deflate",
})

# Resolved lazily on first use
_file_list_url: str = ""
_file_details_url: str = ""
_category_list_url: str = ""


def _init_urls() -> None:
    global _file_list_url, _file_details_url, _category_list_url
    if _file_list_url:
        return
    gc = _SESSION.get(_GLOBAL_CONFIG_URL, timeout=15).json()
    eso = next((g for g in gc.get("GAMES", []) if g.get("GameID") == _GAME_ID), None)
    if not eso:
        raise RuntimeError("ESO not found in MMOUI global config")
    game_cfg = _SESSION.get(eso["GameConfig"], timeout=15).json()
    feeds = game_cfg.get("APIFeeds", {})
    _file_list_url = feeds["FileList"]
    _file_details_url = feeds["FileDetails"]   # append "{id}.json"
    _category_list_url = feeds["CategoryList"]


# ── Data models ───────────────────────────────────────────────────────────────

@dataclass
class Category:
    cat_id: int
    name: str
    icon_url: str
    file_count: int


@dataclass
class RemoteAddonInfo:
    addon_id: int
    name: str              # UIName — display name
    version: str
    date: int              # UIDate unix ms timestamp
    author: str
    description: str       # filled by fetch_addon_details()
    changelog: str         # filled by fetch_addon_details()
    category_id: int
    category: str          # display name, resolved from category list
    downloads_total: int
    downloads_monthly: int
    favorites: int
    info_url: str
    download_url: str      # direct CDN zip URL, filled by fetch_addon_details()
    filename: str          # zip filename
    dirs: list[str]        # UIDir — actual folder name(s) extracted from zip
    md5: str
    thumbs: list[str]


def _to_list(value) -> list[str]:
    """Normalise UIDir/UIIMG_Thumbs — API returns str, list, or null."""
    if isinstance(value, list):
        return [v for v in value if v]
    if isinstance(value, str) and value:
        return [value]
    return []


def _strip_bbcode(text: str) -> str:
    """Remove BBCode tags from description text."""
    text = re.sub(r'\[/?[A-Za-z][^\]]*\]', '', text)
    return text.strip()


def _parse_filelist_entry(raw: dict, cats: dict[int, str]) -> RemoteAddonInfo:
    uid = int(raw.get("UID", 0))
    cat_id = int(raw.get("UICATID", 0))
    return RemoteAddonInfo(
        addon_id=uid,
        name=raw.get("UIName", ""),
        version=raw.get("UIVersion", ""),
        date=int(raw.get("UIDate", 0)),
        author=raw.get("UIAuthorName", ""),
        description="",
        changelog="",
        category_id=cat_id,
        category=cats.get(cat_id, ""),
        downloads_total=int(raw.get("UIDownloadTotal", 0)),
        downloads_monthly=int(raw.get("UIDownloadMonthly", 0)),
        favorites=int(raw.get("UIFavoriteTotal", 0)),
        info_url=raw.get("UIFileInfoURL", ""),
        download_url="",
        filename="",
        dirs=_to_list(raw.get("UIDir")),
        md5="",
        thumbs=_to_list(raw.get("UIIMG_Thumbs")),
    )


# ── Public API ────────────────────────────────────────────────────────────────

def fetch_categories() -> list[Category]:
    _init_urls()
    data = _SESSION.get(_category_list_url, timeout=15).json()
    return [
        Category(
            cat_id=int(c["UICATID"]),
            name=c.get("UICATTitle", ""),
            icon_url=c.get("UICATICON", ""),
            file_count=int(c.get("UICATFileCount", 0)),
        )
        for c in data
    ]


def fetch_all_addons() -> list[RemoteAddonInfo]:
    """Fetch the complete addon list (~3000 items) in one request."""
    _init_urls()
    cats = {c.cat_id: c.name for c in fetch_categories()}
    data = _SESSION.get(_file_list_url, timeout=30).json()
    return [_parse_filelist_entry(raw, cats) for raw in data]


def fetch_addon_details(addon_id: int) -> dict:
    """
    Return full details for one addon: download URL, description, changelog, md5.
    Merges into a dict so callers can update their RemoteAddonInfo.
    """
    _init_urls()
    url = f"{_file_details_url}{addon_id}.json"
    data = _SESSION.get(url, timeout=15).json()
    if isinstance(data, list):
        data = data[0] if data else {}
    return {
        "download_url": data.get("UIDownload", ""),
        "filename":     data.get("UIFileName", ""),
        "md5":          data.get("UIMD5", ""),
        "description":  _strip_bbcode(data.get("UIDescription", "")),
        "changelog":    _strip_bbcode(data.get("UIChangeLog", "")),
    }


def download_zip(
    download_url: str,
    dest_path: str,
    progress_callback=None,
) -> None:
    """Stream a zip to dest_path. progress_callback(bytes_done, total) is optional."""
    with _SESSION.get(download_url, stream=True, timeout=60) as resp:
        resp.raise_for_status()
        total = int(resp.headers.get("content-length", 0))
        done = 0
        with open(dest_path, "wb") as f:
            for chunk in resp.iter_content(chunk_size=65536):
                f.write(chunk)
                done += len(chunk)
                if progress_callback:
                    progress_callback(done, total)

File: src-tauri/src/core/esoui.rs
use once_cell::sync::Lazy;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use tokio::sync::OnceCell;

static CLIENT: Lazy<reqwest::Client> = Lazy::new(|| {
    reqwest::Client::builder()
        .user_agent("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36")
        .gzip(true)
        .build()
        .unwrap()
});

#[derive(Debug, Clone)]
struct ApiUrls {
    file_list: String,
    file_details: String,
    category_list: String,
}

static API_URLS: OnceCell<ApiUrls> = OnceCell::const_new();

async fn get_urls() -> Result<&'static ApiUrls, String> {
    API_URLS
        .get_or_try_init(|| async {
            let gc: serde_json::Value = CLIENT
                .get("https://api.mmoui.com/v3/globalconfig.json")
                .timeout(std::time::Duration::from_secs(15))
                .send()
                .await
                .map_err(|e| e.to_string())?
                .json()
                .await
                .map_err(|e| e.to_string())?;

            let eso = gc["GAMES"]
                .as_array()
                .and_then(|g| g.iter().find(|x| x["GameID"] == "ESO"))
                .ok_or("ESO not found in MMOUI global config")?;

            let game_cfg: serde_json::Value = CLIENT
                .get(eso["GameConfig"].as_str().ok_or("missing GameConfig")?)
                .timeout(std::time::Duration::from_secs(15))
                .send()
                .await
                .map_err(|e| e.to_string())?
                .json()
                .await
                .map_err(|e| e.to_string())?;

            let feeds = &game_cfg["APIFeeds"];
            Ok(ApiUrls {
                file_list: feeds["FileList"].as_str().ok_or("missing FileList")?.to_string(),
                file_details: feeds["FileDetails"].as_str().ok_or("missing FileDetails")?.to_string(),
                category_list: feeds["CategoryList"].as_str().ok_or("missing CategoryList")?.to_string(),
            })
        })
        .await
        .map_err(|e: String| e)
}

fn to_str_vec(v: &serde_json::Value) -> Vec<String> {
    match v {
        serde_json::Value::Array(a) => a.iter().filter_map(|x| x.as_str()).filter(|s| !s.is_empty()).map(String::from).collect(),
        serde_json::Value::String(s) if !s.is_empty() => vec![s.clone()],
        _ => vec![],
    }
}

fn strip_bbcode(s: &str) -> String {
    let re = regex::Regex::new(r"\[/?[A-Za-z][^\]]*\]").unwrap();
    re.replace_all(s, "").trim().to_string()
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RemoteAddon {
    pub addon_id: u32,
    pub name: String,
    pub version: String,
    pub date: i64,
    pub author: String,
    pub description: String,
    pub changelog: String,
    pub category_id: u32,
    pub category: String,
    pub downloads_total: u32,
    pub downloads_monthly: u32,
    pub favorites: u32,
    pub info_url: String,
    pub download_url: String,
    pub filename: String,
    pub dirs: Vec<String>,
    pub md5: String,
}

pub async fn fetch_all_addons() -> Result<Vec<RemoteAddon>, String> {
    let urls = get_urls().await?;

    let cats_raw: Vec<serde_json::Value> = CLIENT
        .get(&urls.category_list)
        .timeout(std::time::Duration::from_secs(15))
        .send()
        .await
        .map_err(|e| e.to_string())?
        .json()
        .await
        .map_err(|e| e.to_string())?;

    let cats: HashMap<u64, String> = cats_raw.iter().filter_map(|c| {
        let id = c["UICATID"].as_str()?.parse::<u64>().ok()?;
        let name = c["UICATTitle"].as_str()?.to_string();
        Some((id, name))
    }).collect();

    let list: Vec<serde_json::Value> = CLIENT
        .get(&urls.file_list)
        .timeout(std::time::Duration::from_secs(30))
        .send()
        .await
        .map_err(|e| e.to_string())?
        .json()
        .await
        .map_err(|e| e.to_string())?;

    Ok(list.iter().map(|raw| {
        let cat_id = raw["UICATID"].as_str().and_then(|s| s.parse::<u32>().ok()).unwrap_or(0);
        RemoteAddon {
            addon_id: raw["UID"].as_str().and_then(|s| s.parse().ok()).unwrap_or(0),
            name: raw["UIName"].as_str().unwrap_or("").to_string(),
            version: raw["UIVersion"].as_str().unwrap_or("").to_string(),
            date: raw["UIDate"].as_str().and_then(|s| s.parse().ok()).unwrap_or(0),
            author: raw["UIAuthorName"].as_str().unwrap_or("").to_string(),
            description: String::new(),
            changelog: String::new(),
            category_id: cat_id,
            category: cats.get(&(cat_id as u64)).cloned().unwrap_or_default(),
            downloads_total: raw["UIDownloadTotal"].as_str().and_then(|s| s.parse().ok()).unwrap_or(0),
            downloads_monthly: raw["UIDownloadMonthly"].as_str().and_then(|s| s.parse().ok()).unwrap_or(0),
            favorites: raw["UIFavoriteTotal"].as_str().and_then(|s| s.parse().ok()).unwrap_or(0),
            info_url: raw["UIFileInfoURL"].as_str().unwrap_or("").to_string(),
            download_url: String::new(),
            filename: String::new(),
            dirs: to_str_vec(&raw["UIDir"]),
            md5: String::new(),
        }
    }).collect())
}

#[derive(Debug, Serialize, Deserialize)]
pub struct AddonDetails {
    pub download_url: String,
    pub filename: String,
    pub md5: String,
    pub description: String,
    pub changelog: String,
}

pub async fn fetch_addon_details(addon_id: u32) -> Result<AddonDetails, String> {
    let urls = get_urls().await?;
    let url = format!("{}{}.json", urls.file_details, addon_id);
    let raw: serde_json::Value = CLIENT
        .get(&url)
        .timeout(std::time::Duration::from_secs(15))
        .send()
        .await
        .map_err(|e| e.to_string())?
        .json()
        .await
        .map_err(|e| e.to_string())?;

    let data = if raw.is_array() { raw[0].clone() } else { raw };
    Ok(AddonDetails {
        download_url: data["UIDownload"].as_str().unwrap_or("").to_string(),
        filename: data["UIFileName"].as_str().unwrap_or("").to_string(),
        md5: data["UIMD5"].as_str().unwrap_or("").to_string(),
        description: strip_bbcode(data["UIDescription"].as_str().unwrap_or("")),
        changelog: strip_bbcode(data["UIChangeLog"].as_str().unwrap_or("")),
    })
}

pub async fn download_zip(
    url: &str,
    dest: &std::path::Path,
    on_progress: impl Fn(u64, u64),
) -> Result<(), String> {
    use tokio::io::AsyncWriteExt;
    let mut resp = CLIENT
        .get(url)
        .timeout(std::time::Duration::from_secs(120))
        .send()
        .await
        .map_err(|e| e.to_string())?;

    let total = resp.content_length().unwrap_or(0);
    let mut file = tokio::fs::File::create(dest).await.map_err(|e| e.to_string())?;
    let mut done = 0u64;

    while let Some(chunk) = resp.chunk().await.map_err(|e| e.to_string())? {
        file.write_all(&chunk).await.map_err(|e| e.to_string())?;
        done += chunk.len() as u64;
        on_progress(done, total);
    }
    Ok(())
}
